src/extension.ts

```diff
@@ -869,44 +884,60 @@ function findPod(callback) {
     );
 }
 
-function selectPodForApp(callback) {
-    findPodsForApp((podList) => {
-        if (podList.items.length === 0) {
-            vscode.window.showErrorMessage('Couldn\'t find any relevant pods.');
-            callback(null);
-            return;
-        }
-        if (podList.items.length === 1) {
-            callback(podList.items[0]);
-            return;
-        }
-        let names = [];
+enum PodSelectionFallback {
+    None,
+    AnyPod,
+}
 
-        for (const element of podList) {
-            names.push(`${element.metadata.namespace}/${element.metadata.name}`);
+enum PodSelectionScope {
+    App,
+    All,
+}
+
+async function selectPod(scope: PodSelectionScope, fallback: PodSelectionFallback) : Promise<any | null> {
+    const findPodsResult = scope === PodSelectionScope.App ? await findPodsForApp() : await findAllPods();
+    
+    if (!findPodsResult.succeeded) {
+        return null;
+    }
+
+    const podList = findPodsResult.pods;
+
+    if (podList.length === 0) {
+        if (fallback === PodSelectionFallback.AnyPod) {
+            return selectPod(PodSelectionScope.All, PodSelectionFallback.None);
         }
+        const scopeMessage = scope === PodSelectionScope.App ? "associated with this app" : "in the cluster";
+        vscode.window.showErrorMessage(`Couldn't find any pods ${scopeMessage}.`);
+        return null;
+    }
+    if (podList.length === 1) {
+        return podList[0];
+    }
+    let names = [];
 
-        vscode.window.showQuickPick(names).then((value) => {
-            if (!value) {
-                callback(null);
-                return;
-            }
+    for (const element of podList) {
+        names.push(`${element.metadata.namespace}/${element.metadata.name}`);
+    }
 
-            let ix = value.indexOf('/');
-            let name = value.substring(ix + 1);
+    const value = await vscode.window.showQuickPick(names);
```

Comment:
Suggest to use `vscode.window.showQuickPick<T extends vscode.QuickPickItem>(items: T[])`, no need to loop the array again to query the original element after picking. 

See the sample https://github.com/Azure/vscode-kubernetes-tools/blob/master/src/debug/providerRegistry.ts#L11